README.md
# Font Awesome for AngularDart

## Basic Usage

This package provides Font Awesome as an AngularDart component. To use icons
inside a component, do the following:

* Import the package: `import 'package:ng_fontawesome/cng_fontawesome.dart';`
* Add the directives to your component: `directives: const [FA_DIRECTIVES]`
* Add icons to your template, e.g. `<fa name="flag"></fa>`

See more usage information in the [demo
application](https://hyperiongray.github.io/ng_fontawesome_demo/) ([source 
code](https://github.com/hyperiongray/ng_fontawesome_demo)).

## Versioning

The version number of this package follows the version number of Font Awesome
itself. This makes it easy to see which version of Font Awesome is provided by
this package and pin your own packages to specific versions of Font Awesome
(i.e. a version that adds a specific glyph that your package needs).

## Maintenance

_This section is for the developer[s] of this package only._

To upgrade Font Awesome, download and unzip the upstream release. Change into
the unzipped directory. Then run these commands:

    $ cp css/all.* ../lib/src/css
    $ cp webfonts/* ../lib/fonts
    $ cd ..
    $ sed --in-place 's:url[(]"\.\./web:url("packages/ng_fontawesome/:g' lib/src/css/all.css
    $ sed --in-place 's:url[(]\.\./web:url(packages/ng_fontawesome/:g' lib/src/css/all.min.css

To test it out, switch to the `ng_fontawesome_demo` project and run:

    $ pub get
    $ webdev serve

To publish:

1. Put new version in `pubspec.yaml`
2. Update `CHANGELOG.md`
3. Run `dartdoc` to make sure it can run successfully
4. Create and push a git tag
5. Publish: `pub publish`

---

[![define hyperion gray](https://hyperiongray.s3.amazonaws.com/define-hg.svg)](https://www.hyperiongray.com/?pk_campaign=github&pk_kwd=ng_fontawesome "Hyperion Gray")


CHANGELOG.md
The package's version number follows the upstream Font Awesome version number.
The build number (after the plus sign) is increased when the package is modified
without updating the upstream.

5.7.2+1
=======

- Upgrade to Dart 2.x
- Upgrade to Angular 5.x
- Upgrade to Font Awesome 5.7.2

4.7.0+6
=======

- Improve Pub score

4.7.0+5
=======

- Renamed from ng2_fontawesome to ng_fontawesome
- Upgraded to Angular 4.0.0



lib/src/icon.dart
import 'package:angular/angular.dart';

/// A component that renders a solid icon from Font Awesome.
@Component(
    selector: 'fa',
    styleUrls: const ['css/all.min.css'],
    template: '''<i [ngClass]='classes'></i>''',
    directives: const [coreDirectives]
)
class FaIcon implements OnInit {

    /// Name of icon to dislpay, e.g. "flag" or "camera-retro". (Required)
    @Input()
    String name;

    /// The group of icon: regular, solid, light, or brand. (Default: solid)
    @Input()
    String group = 'solid';

    static const GROUP_CLASSES = const {
        'brand': 'fab',
        'regular': 'far',
        'solid': 'fas',
    };

    /// Size of icon: xs, sm, lg, 2x, 3x, 5x, 7x, or 10x.
    @Input()
    String size;

    /// Render with fixed width.
    @Input()
    bool fw = false;

    /// Rotate an icon: 90, 180, or 270.
    @Input()
    String rotate;

    static const ROTATE_CLASSES = const {
        '90': 'fa-rotate-90',
        '180': 'fa-rotate-180',
        '270': 'fa-rotate-270',
    };

    /// Flip an icon: horizontal, vertical, or both.
    @Input()
    String flip;

    static const FLIP_CLASSES = const {
        'horizontal': 'fa-flip-horizontal',
        'vertical': 'fa-flip-vertical',
        'both': 'fa-flip-both',
    };

    /// Animate a rotating icon smoothly.
    @Input()
    bool spin = false;

    /// Animate a rotating icon in 8 discrete steps.
    @Input()
    bool pulse = false;

    /// Float "left" or "right".
    @Input()
    String pull;

    static const PULL_CLASSES = const {
        'left': 'fa-pull-left',
        'right': 'fa-pull-right',
    };

    /// Render with border around the icon.
    @Input()
    bool border = false;

    /// Invert color scheme. Only meaningful when the component appears inside
    /// an <fa-stack>.
    @Input()
    bool inverse = false;

    /// The stacking size, either "1x" or "2x".
    ///
    /// Only meaningful when the component appears inside an <fa-stack>.
    @Input()
    String stack;

    static const STACK_CLASSES = const {
        '1x': 'fa-stack-1x',
        '2x': 'fa-stack-2x',
    };

    Map<String, bool> classes = {};

    void ngOnInit() {
        var group = FaIcon.GROUP_CLASSES[this.group];
        this.classes['fa-$name'] = true;
        this.classes[group] = true;
        if (this.size != null) {
            this.classes['fa-$size'] = true;
        }
        this.classes['fa-fw'] = this.fw;
        if (this.rotate != null) {
            var rotate = FaIcon.ROTATE_CLASSES[this.rotate];
            this.classes[rotate] = true;
        }
        if (this.flip != null) {
            var flip = FaIcon.FLIP_CLASSES[this.flip];
            this.classes[flip] = true;
        }
        this.classes['fa-spin'] = this.spin;
        this.classes['fa-pulse'] = this.pulse;
        if (this.pull != null) {
            var pull = FaIcon.PULL_CLASSES[this.pull];
            this.classes[pull] = true;
        }
        this.classes['fa-border'] = this.border;
        this.classes['fa-inverse'] = this.inverse;
        if (this.stack != null) {
            var stack = FaIcon.STACK_CLASSES[this.stack];
            this.classes[stack] = true;
        }
    }
}


lib/src/list.dart
import 'package:angular/core.dart';

import 'icon.dart';

/// A component that renders an unordered list with Font Awesome icons as
/// bullets.
@Component(
    selector: 'fa-ul',
    styleUrls: const ['css/all.min.css'],
    template: '''<ul style="position: relative" class="fa-ul">
        <ng-content></ng-content></ul>'''
)
class FaList implements AfterContentInit {
    /// Default bullet to use for list items.
    @Input()
    String bullet;

    /// The group of the default bullet: solid, regular, or brand. (Default:
    /// solid)
    @Input()
    String group = 'solid';

    /// Icon list items nested inside this element.
    @ContentChildren(FaListItem)
    List<FaListItem> childIcons;

    /// Implementation of AfterContentInit.
    void ngAfterContentInit() {
        if (this.bullet != null) {
            _copyBulletToChildren();
        }
    }

    /// Copy the bullet icon into any children that don't have one.
    void _copyBulletToChildren() {
        for (var childIcon in childIcons) {
            if (childIcon.bullet == null) {
                childIcon.bullet = this.bullet;
                childIcon.group = this.group;
            }
        }
    }
}

/// A component that renders a list item with a Font Awesome icon as the
/// bullet.
@Component(
    selector: 'fa-li',
    styleUrls: const ['css/all.min.css'],
    template: '''<li><span class="fa-li"><fa [group]="group" [name]="bullet">
        </fa></span><ng-content></ng-content></li>''',
    directives: const [FaIcon]
)
class FaListItem {
    /// Name of icon to use for this list item's bullet.
    @Input()
    String bullet;

    /// The group of the bullet: solid, regular, or brand. (Default: solid)
    @Input()
    String group = 'solid';
}


lib/ng2_fontawesome.dart
import 'package:angular2/core.dart';

/// A component that renders an icon from Font Awesome.
@Component(
    selector: 'fa',
    styleUrls: const ['css/font-awesome.min.css'],
    template: '''<i class='fa fa-{{name}}'
        [class.fa-border]='border'
        [class.fa-fw]='fw'
        [class.fa-flip-horizontal]='flip == "horizontal"'
        [class.fa-flip-vertical]='flip == "vertical"'
        [class.fa-inverse]='inverse'
        [class.fa-li]
        [class.fa-pull-left]='pull == "left"'
        [class.fa-pull-right]='pull == "right"'
        [class.fa-pulse]='pulse'
        [class.fa-rotate-90]='rotate == 90'
        [class.fa-rotate-180]='rotate == 180'
        [class.fa-rotate-270]='rotate == 270'
        [class.fa-lg]='size == "lg"'
        [class.fa-2x]='size == "2x"'
        [class.fa-3x]='size == "3x"'
        [class.fa-4x]='size == "4x"'
        [class.fa-5x]='size == "5x"'
        [class.fa-spin]='spin'
        [class.fa-stack-1x]='stack == "1x"'
        [class.fa-stack-2x]='stack == "2x"'
      ></i>'''
)
class FaIcon {
    /// Render with border around the icon.
    @Input()
    bool border = false;

    /// Flip along "horizontal" or "vertical" axis.
    @Input()
    String flip;

    /// Render with fixed width.
    @Input()
    bool fw = false;

    /// Invert color scheme.
    @Input()
    bool inverse = false;

    /// Name of icon to dislpay, e.g. "flag" or "camera-retro". (Required)
    @Input()
    String name;

    /// Float "left" or "right".
    @Input()
    String pull;

    /// Animate a rotating icon in 8 discrete steps.
    @Input()
    bool pulse = false;

    /// Orient the icon in a different direction: 90, 180, OR 270 degrees
    /// clockwise.
    @Input()
    int rotate;

    /// Size of the icon. "lg" is 33% larger than normal. Other options are
    /// "2x", "3x", "4x", and "5x".
    @Input()
    String size;

    /// Animate a rotating icon smoothly.
    @Input()
    bool spin = false;

    /// The stacking size, either "1x" or "2x".
    ///
    /// Only useful when the component appears inside an <fa-stack>.
    @Input()
    String stack;
}

/// A component that renders an unordered list with Font Awesome icons as
/// bullets.
@Component(
    selector: 'fa-ul',
    styleUrls: const ['css/font-awesome.min.css'],
    template: '<ul class="fa-ul"><ng-content></ng-content></ul>'
)
class FaList implements AfterContentInit {
    /// Default bullet to use for list items.
    @Input()
    String bullet;

    @ContentChildren(FaListItem)
    QueryList<FaListItem> childIcons;

    /// Implementation of AfterContentInit.
    void ngAfterContentInit() {
        if (bullet != null) {
            _copyBulletToChildren();
        }
    }

    /// Copy the bullet icon into any children that don't have one.
    void _copyBulletToChildren() {
        for (var childIcon in childIcons) {
            if (childIcon.bullet == null) {
                childIcon.bullet = bullet;
            }
        }
    }
}

/// A component that renders a list item with a Font Awesome icons as the
/// bullet.
@Component(
    selector: 'fa-li',
    styleUrls: const ['css/font-awesome.min.css'],
    template: '<li><fa [name]="bullet"></fa><ng-content></ng-content></li>',
    directives: const [FaIcon]
)
class FaListItem {
    /// Name of icon to use for this list item's bullet.
    @Input()
    String bullet;
}

/// A component that composites Font Awesome icons on top of each other.
@Component(
    selector: 'fa-stack',
    styleUrls: const ['css/font-awesome.min.css'],
    template: '<span class="fa-stack"><ng-content></ng-content></span>',
    directives: const [FaIcon]
)
class FaStack {}

const List<dynamic> FA_DIRECTIVES = const [
    FaIcon,
    FaList,
    FaListItem,
    FaStack,
];
